### PR TITLE
[flang] Added extension point callbacks to default FIR optimizer pipeline.

### PR DESCRIPTION
**Pull request by VijayKandiah**
This change inserts a few extension point callbacks in the DefaultFIROptimizerPassPipeline. As an example usage of callbacks in the FIR optimizer pipeline,  the FIRInlinerCallback is now used to register the default MLIR inliner pass in flang-new, tco, and bbc compilation flows. Other compilation flows can use these callbacks to add extra passes at different points of the pass pipeline.

### CODE REVIEW

**Review by clementval — APPROVED** [review]
LGTM. Looks like you need to run clang-format on your change.

File: flang/lib/Frontend/FrontendActions.cpp [hunk 1]
```diff
@@ -802,6 +802,7 @@ void CodeGenAction::generateLLVMIR() {
   pm.enableVerifier(/*verifyPasses=*/true);
 
   MLIRToLLVMPassPipelineConfig config(level, opts, mathOpts);
+  fir::registerDefaultInlinerPass(config);
```
Comment (kiranchandramohan) [inline flang/lib/Frontend/FrontendActions.cpp:805]: Is the pass always registered and executed (irrespective of the optimization level) with the flang-driver? Is that expected and different from tco?

**Review by kiranchandramohan — APPROVED** [review]
Thanks Vijay for the patch. I have a couple of minor questions inline. 

It might be good to document the avaliability of these callbacks in flang/docs. Is this intended for Drivers (upstream/downstream) to register callbacks statically? Or is the plan to allow drivers to load FIR/MLIR passes dynamically with this feature?

**Comment by VijayKandiah** [comment]
Thanks Kiran for the feedback. Could you please clarify what you mean by statically vs dynamically registering callbacks? I plan to document these callbacks [here](https://github.com/llvm/llvm-project/blob/main/flang/docs/FlangDriver.md) with the next commit before merging.

File: flang/lib/Frontend/FrontendActions.cpp (hunk 1 above)
Comment (VijayKandiah) [reply flang/lib/Frontend/FrontendActions.cpp:805]: Yes, that was the original behavior to register and execute the inliner pass regardless of optimization level with flang-new, tco, and bbc. The changes preserve this behavior.

File: flang/include/flang/Tools/CLOptions.inc [hunk 2]
```diff
@@ -232,12 +232,27 @@ inline void addExternalNameConversionPass(
   });
 }
 
+// Use inliner extension point callback to register the default inliner pass.
+inline void registerDefaultInlinerPass(MLIRToLLVMPassPipelineConfig &config) {
+  config.registerFIRInlinerCallback(
+      [](mlir::PassManager &pm, llvm::OptimizationLevel level) {
+        llvm::StringMap<mlir::OpPassManager> pipelines;
+        // The default inliner pass adds the canonicalizer pass with the default
+        // configuration. Create the inliner pass with tco config.
```
[the remark this answers is not in the document]
Comment (VijayKandiah) [reply flang/include/flang/Tools/CLOptions.inc]: Thanks for pointing this out, I dropped the line in my latest commit.

**Comment by kiranchandramohan** [comment]
> Thanks Kiran for the feedback. Could you please clarify what you mean by statically vs dynamically registering callbacks?

By dynamic I meant something like what is done with the pass-plugin option. This allows to load a pass from a shared object file during compilation and need not be statically listed in the upstream or downstream source code of the compiler.
https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fpass-plugin
https://medium.com/@mshockwave/writing-pass-instrument-for-llvm-newpm-f17c57d3369f
https://reviews.llvm.org/D129156

File: flang/lib/Frontend/FrontendActions.cpp (hunk 1 above)
Comment (kiranchandramohan) [reply flang/lib/Frontend/FrontendActions.cpp:805]: I was confused by the reference to the O2 optimizer pipelines in both bbc and tco. Thanks for clarifying.

**Comment by VijayKandiah** [comment]
Ah I see, thanks for clarifying. These callbacks currently provide a static way to extend the pass pipeline by upstream or downstream driver source code. However, I see this change as a step in the right direction towards having a dynamic way to load passes with something like a FIR/MLIR pass-plugin option that uses these callbacks to do so.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Pull request merged**